File: reference/spring-boot-java-maven-layered/src/main/java/com/example/demo/service/WidgetService.java
package com.example.demo.service;

import com.example.demo.domain.Widget;
import com.example.demo.repository.WidgetRepository;
import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Business rules live here, not in the controller and not in the entity. The controller maps
 * HTTP to method calls; this class decides what is allowed.
 */
@Service
@Transactional(readOnly = true)
public class WidgetService {

    private final WidgetRepository widgets;

    public WidgetService(WidgetRepository widgets) {
        this.widgets = widgets;
    }

    public List<Widget> findAll() {
        return widgets.findAll();
    }

    public Widget findById(long id) {
        return widgets.findById(id).orElseThrow(() -> new WidgetNotFoundException(id));
    }

    @Transactional
    public Widget create(String name, int quantity) {
        if (widgets.existsByName(name)) {
            throw new DuplicateWidgetNameException(name);
        }
        return widgets.save(Widget.of(name, quantity));
    }

    @Transactional
    public Widget restock(long id, int amount) {
        Widget widget = findById(id);
        widget.restock(amount);
        return widget;
    }

    @Transactional
    public void delete(long id) {
        if (!widgets.existsById(id)) {
            throw new WidgetNotFoundException(id);
        }
        widgets.deleteById(id);
    }
}

File: reference/spring-boot-kotlin-gradle-layered/src/main/kotlin/com/example/demo/service/WidgetService.kt
package com.example.demo.service

import com.example.demo.domain.Widget
import com.example.demo.repository.WidgetRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * Business rules live here, not in the controller and not in the entity. The controller maps
 * HTTP to method calls; this class decides what is allowed.
 */
@Service
@Transactional(readOnly = true)
class WidgetService(
    private val widgets: WidgetRepository,
) {
    fun findAll(): List<Widget> = widgets.findAll()

    fun findById(id: Long): Widget = widgets.findById(id).orElseThrow { WidgetNotFoundException(id) }

    @Transactional
    fun create(
        name: String,
        quantity: Int,
    ): Widget {
        if (widgets.existsByName(name)) {
            throw DuplicateWidgetNameException(name)
        }
        return widgets.save(Widget.of(name, quantity))
    }

    @Transactional
    fun restock(
        id: Long,
        amount: Int,
    ): Widget {
        val widget = findById(id)
        widget.restock(amount)
        return widget
    }

    @Transactional
    fun delete(id: Long) {
        if (!widgets.existsById(id)) {
            throw WidgetNotFoundException(id)
        }
        widgets.deleteById(id)
    }
}
